lib/libutee/arch/arm/ftrace/ftrace.c

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2019, Linaro Limited
+ */
+
+#include <user_ta_header.h>
+#include <utee_syscalls.h>
+#include "ftrace.h"
+
+static const char hex_str[] = "0123456789abcdef";
+
+/*
+ * This api shifts/moves ftrace buffer to create space for new dump
+ * in case the buffer size falls short of actual dump.
+ */
+static void __noprof fbuf_shift(struct ftrace_buf *fbuf, size_t size)
+{
+	char *dst = (char *)fbuf + fbuf->buf_off;
+	const char *src = (char *)fbuf + fbuf->buf_off + size;
+	size_t n = 0;
+
+	fbuf->curr_size -= size;
+
+	for (n = 0; n < fbuf->curr_size; n++)
+		dst[n] = src[n];
+}
+
+static size_t __noprof to_func_enter_fmt(char *buf, uint32_t ret_idx,
+					 unsigned long pc)
+{
+	char *str = buf;
+	uint32_t addr_size = 2 * sizeof(unsigned long);
+	uint32_t i = 0;
+
+	for (i = 0; i < ret_idx; i++)
+		*str++ = ' ';
+
+	*str++ = '0';
+	*str++ = 'x';
+
+	for (i = 0; i < addr_size; i++)
+		*str++ = hex_str[(pc >> 4 * (addr_size - i - 1)) & 0xf];
+
+	*str++ = '(';
+	*str++ = ')';
+	*str++ = ' ';
+	*str++ = '{';
+	*str++ = '\n';
+	*str = '\0';
+
+	return str - buf;
+}
+
+void __noprof ftrace_enter(unsigned long pc, unsigned long *lr)
+{
+	struct ftrace_buf *fbuf = NULL;
+	size_t dump_size = 0;
+
+	fbuf = &__ftrace_buf_start;
+
+	if (!fbuf->buf_off || !fbuf->max_size)
+		return;
+
+	dump_size = fbuf->ret_idx + (2 * sizeof(unsigned long)) + 8;
+
+	if ((fbuf->curr_size + dump_size) > fbuf->max_size)
```

Comment:
It seems we've reached the capacity of the ftrace buffer and need to throw away the oldest data under the assumption that that's the least interesting data. Please add a comment explaining this.

By the way, can the tools we have cope with a removed or broken ingress of the trace buffer?

Reply:
Ok, will add a comment. And yes `symbolize.py` script takes care of broken starting line of dump.